syft/pkg/collection.go

```diff
@@ -283,6 +285,8 @@ func (c *Collection) Enumerate(types ...Type) <-chan Package {
 // is specified.
 func (c *Collection) Sorted(types ...Type) (pkgs []Package) {
 	for p := range c.Enumerate(types...) {
+		// sort CPE for each package so SBOM have deterministic output
+		sort.Sort(cpe.BySourceThenSpecificity(p.CPEs))
```

Comment:
we probably shouldn't be mutating these on read from the collection, but instead at generation of this list. Probably one of these two places:
- https://github.com/anchore/syft/blob/13d01ecaff0aae22f3a5bf6f922a0be4ae7c4d06/syft/pkg/cataloger/internal/cpegenerate/generate.go#L172
- https://github.com/anchore/syft/blob/13d01ecaff0aae22f3a5bf6f922a0be4ae7c4d06/syft/pkg/cataloger/common/cpe/create.go#L10

Reply:
@wagoodman changes made - similar test output showing only two variants that stem from the java cataloger non determinism 
```
./sha.fish
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_1.json\ttest_1.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_2.json\ttest_2.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_3.json\ttest_3.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_4.json\ttest_4.json
5d5655ad32cdbfbcdaaabe6a314f6656145ad33dcb54428802fd3dc153e436d0  test_5.json\ttest_5.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_6.json\ttest_6.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_7.json\ttest_7.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_8.json\ttest_8.json
5d5655ad32cdbfbcdaaabe6a314f6656145ad33dcb54428802fd3dc153e436d0  test_9.json\ttest_9.json
4d076f84e08b338743ee76f68f2ba516267d954dece28ff700c6ebc152ca95a4  test_10.json\ttest_10.json
```